sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobAsyncClientBase.java

```diff
@@ -767,29 +745,201 @@ private AsynchronousFileChannel downloadToFileResourceSupplier(String filePath)
+                 */
+                long newCount = r.getCount() == null || r.getCount() > (totalLength - r.getOffset())
+                    ? totalLength - r.getOffset() : r.getCount();
+
+                return Mono.zip(Mono.just(newCount), Mono.just(newConditions), Mono.just(response));
+            })
+            .onErrorResume(throwable -> {
+                /*
+                In the case of an empty blob, we still want to report success and give back valid headers. Attempting a
+                range download on an empty blob will return an InvalidRange error code and a Content-Range header of the
+                format "bytes * /0". We need to double check that the total size is zero in the case that the customer
+                has attempted an invalid range on a non-zero length blob.
+                 */
+                if (throwable instanceof BlobStorageException
+                    && ((BlobStorageException) throwable).getErrorCode() == BlobErrorCode.INVALID_RANGE
+                    && extractTotalBlobLength(((BlobStorageException) throwable).getResponse()
+                    .getHeaders().getValue("Content-Range")) == 0) {
+
+                    return this.downloadWithResponse(new BlobRange(0, 0L), downloadRetryOptions, requestConditions,
+                        rangeGetContentMd5, context)
+                        .subscribeOn(Schedulers.elastic())
+                        .flatMap(response -> {
+                            /*
+                            Ensure the blob is still 0 length by checking our download was the full length.
+                            (200 is for full blob; 206 is partial).
+                             */
+                            if (response.getStatusCode() != 200) {
+                                Mono.error(new IllegalStateException("Blob was modified mid download. It was "
+                                    + "originally 0 bytes and is now larger."));
+                            }
+                            return Mono.zip(Mono.just(0L), Mono.just(requestConditions), Mono.just(response));
+                        });
+                }
+                return Mono.error(throwable);
+            });
+    }
+
+    private static BlobRequestConditions setEtag(BlobRequestConditions accessConditions, String etag) {
+        //We don't want to modify the user's object, so we'll create a duplicate and set the retrieved etag.
+        return new BlobRequestConditions()
+            .setIfModifiedSince(
+                accessConditions.getIfModifiedSince())
+            .setIfUnmodifiedSince(
+                accessConditions.getIfModifiedSince())
+            .setIfMatch(etag)
+            .setIfNoneMatch(
+                accessConditions.getIfNoneMatch())
+            .setLeaseId(accessConditions.getLeaseId());
+    }
+
+    private static Mono<Response<BlobProperties>> writeBodyToFile(BlobDownloadAsyncResponse response,
```

Comment:
Could we change this to be `Mono<Void>` and change the `response` to `Flux<ByteBuffer> data` and then in the stream that calls into this instead of using `.last()` we use `thenReturn(<new method that creates the response>)`, that way we aren't constructing a new response object during every write being performed.